File: Algorithm/Python/1000+/1043. Partition Array for Maximum Sum.py
class Solution(object):
    def maxSumAfterPartitioning(self, arr, k):
        """
        :type arr: List[int]
        :type k: int
        :rtype: int
        """
        n = len(arr)
        dp = [0] * (n+1)
        for i in range(n):
            max_num, cnt = arr[i], 0
            while cnt < k and i-cnt >= 0:
                max_num = max(max_num, arr[i-cnt])
                dp[i+1] = max(dp[i-cnt] + max_num*(cnt+1), dp[i+1])
                cnt += 1
        return dp[-1]

File: Algorithm/C++/1000+/1043. Partition Array for Maximum Sum.cc
#include <vector>
#include <algorithm>

class Solution {
public:
    int maxSumAfterPartitioning(std::vector<int>& arr, int k) {
        int n = arr.size();
        std::vector<int> dp(n+1, 0);
        for (int i = 0; i < n; i++) {
            int offset = 0, _max = arr[i];
            while (offset < k && i-offset >= 0) {
                _max = std::max(_max, arr[i-offset]);
                dp[i+1] = std::max(dp[i+1], dp[i-offset] + _max*(offset+1));
                offset++;
            }
        }
        return dp[n];
    }
};